server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java

```diff
@@ -8,77 +8,131 @@
+import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.util.concurrent.DeterministicTaskQueue;
+import org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor;
 import org.elasticsearch.core.TimeValue;
-import org.elasticsearch.index.IndexVersion;
-import org.elasticsearch.tasks.TaskManager;
+import org.elasticsearch.test.ClusterServiceUtils;
 import org.elasticsearch.test.ESTestCase;
-import org.hamcrest.Matchers;
+import org.junit.Before;
 
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.common.settings.ClusterSettings.createBuiltInClusterSettings;
+import static org.elasticsearch.gateway.GatewayService.EXPECTED_DATA_NODES_SETTING;
+import static org.elasticsearch.gateway.GatewayService.RECOVER_AFTER_DATA_NODES_SETTING;
+import static org.elasticsearch.gateway.GatewayService.RECOVER_AFTER_TIME_SETTING;
 import static org.elasticsearch.gateway.GatewayService.STATE_NOT_RECOVERED_BLOCK;
-import static org.elasticsearch.test.NodeRoles.masterNode;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.hamcrest.CoreMatchers.sameInstance;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 import static org.hamcrest.Matchers.hasItem;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 public class GatewayServiceTests extends ESTestCase {
 
-    private GatewayService createService(final Settings.Builder settings) {
-        final ClusterService clusterService = new ClusterService(
-            Settings.builder().put("cluster.name", "GatewayServiceTests").build(),
-            new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS),
-            null,
-            (TaskManager) null
+    private DeterministicTaskQueue deterministicTaskQueue;
+    private AtomicInteger rerouteCount;
+    private ClusterService clusterService;
+    private String dataNodeIdPrefix;
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        deterministicTaskQueue = new DeterministicTaskQueue();
+        assertThat(deterministicTaskQueue.getCurrentTimeMillis(), equalTo(0L));
+        rerouteCount = new AtomicInteger();
+        dataNodeIdPrefix = randomAlphaOfLength(10) + "-";
+    }
+
+    private GatewayService createService(final Settings.Builder settingsBuilder, final ClusterState initialState) {
+        final var threadPool = deterministicTaskQueue.getThreadPool();
+        final var settings = settingsBuilder.build();
+        final var clusterSettings = createBuiltInClusterSettings(settings);
+
+        clusterService = new ClusterService(
```

Comment:
Sure I pushed b5da2d1edcb1c5dd617cbf5ee191e09007cb7690

An easier alternative is to access `ClusterService` from `GatewayService`. But it requires making the `GatewayService.clusterService` field package private (or add a package private accessor method). Though this pattern is used in many places for testing, I am not sure whether you'd like it. So I did not take this approach. Please let me know whether this would work for you since it might be useful in future.